lib/googleauth/credentials.rb

```diff
@@ -0,0 +1,156 @@
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+require 'forwardable'
+require 'json'
+require 'signet/oauth_2/client'
+
+require 'googleauth/default_credentials'
+
+module Google
+  module Auth
+    # This class is intended to be inherited by API-specific classes
+    # which overrides the SCOPE constant.
+    class Credentials
+      TOKEN_CREDENTIAL_URI = 'https://accounts.google.com/o/oauth2/token'.freeze
+      AUDIENCE = 'https://accounts.google.com/o/oauth2/token'.freeze
+      SCOPE = [].freeze
+      PATH_ENV_VARS = [].freeze
+      JSON_ENV_VARS = [].freeze
+      DEFAULT_PATHS = [].freeze
+
+      attr_accessor :client
+
+      # Delegate client methods to the client object.
+      extend Forwardable
+      def_delegators :@client,
+                     :token_credential_uri, :audience,
+                     :scope, :issuer, :signing_key, :updater_proc
+
+      def initialize(keyfile, options = {})
+        scope = options[:scope]
+        verify_keyfile_provided! keyfile
+        if keyfile.is_a? Signet::OAuth2::Client
+          @client = keyfile
+        elsif keyfile.is_a? Hash
+          hash = stringify_hash_keys keyfile
+          hash['scope'] ||= scope
+          @client = init_client hash
+        else
+          verify_keyfile_exists! keyfile
+          json = JSON.parse ::File.read(keyfile)
+          json['scope'] ||= scope
+          @client = init_client json
+        end
+        @client.fetch_access_token!
+      end
+
+      # Returns the default credentials.
+      def self.default(options = {})
+        scope = options[:scope]
+        # First try to find keyfile file from environment variables.
+        self::PATH_ENV_VARS.map(&method(:env))
+                           .compact
+                           .select(&method(:path))
+                           .each do |file|
+          return new file, scope: scope
+        end
+        # Second try to find keyfile json from environment variables.
+        self::JSON_ENV_VARS.map(&method(:json)).compact.each do |hash|
+          return new hash, scope: scope
+        end
+        # Third try to find keyfile file from known file paths.
+        self::DEFAULT_PATHS.select(&method(:path)).each do |file|
+          return new file, scope: scope
+        end
+        # Finally get instantiated client from Google::Auth.
+        scope ||= self::SCOPE
+        client = Google::Auth.get_application_default scope
+        new client
+      end
+
+      def self.env(v)
```

Comment:
These are awkward as class methods. Instead, consider using `Kernel#lambda` to make the original `->()` expressions 1.9-compatible. For example:

```ruby
env = lambda{ |v| ENV[v] }
# etc.
```

Reply:
Done.